packages/next/vitals/vitals.ts

```diff
@@ -0,0 +1,29 @@
+import { useEffect, useRef } from 'react'
+import { NextWebVitalsMetric } from '../pages/_app'
+
+type ReportWebVitalsCallback = (webVitals: NextWebVitalsMetric) => any
+export const webVitalsCallbacks = new Set<ReportWebVitalsCallback>()
+const metrics: NextWebVitalsMetric[] = []
+
+export function trackWebVitalMetric(metric: NextWebVitalsMetric) {
+  metrics.push(metric)
+  webVitalsCallbacks.forEach((callback) => callback(metric))
+}
+
+export function useWebVitalsReport(callback: ReportWebVitalsCallback) {
+  const metricIndexRef = useRef(0)
+
+  useEffect(() => {
+    // Flush calculated metrics
+    const reportMetric = (metric: NextWebVitalsMetric) => callback(metric)
+    for (let i = metricIndexRef.current; i < metrics.length; i++) {
+      reportMetric(metrics[i])
+      metricIndexRef.current = i
+    }
```

Comment:
```suggestion
    for (let i = metricIndexRef.current; i < metrics.length; i++) {
      reportMetric(metrics[i])
    }
    metricIndexRef.current = metrics.length
```

Otherwise `metricIndexRef.current` will end up with `metrics.length - 1`, and when a new effect is triggered it still enters the loop and causes duplicated reports.